skfp/model_selection/splitters/butina_split.py

```diff
@@ -0,0 +1,448 @@
+
+    .. [4] `Roger Sayle
+        "2D similarity, diversity and clustering in RDKit"
+        RDKit UGM 2019
+        <https://www.nextmovesoftware.com/talks/Sayle_2DSimilarityDiversityAndClusteringInRdkit_RDKITUGM_201909.pdf>`_
+
+    .. [5] `W. Dong et al.
+        "Efficient k-nearest neighbor graph construction for generic similarity measures"
+        Proceedings of the 20th International World Wide Web Conference (WWW '11).
+        Association for Computing Machinery, New York, NY, USA, 577–586
+        <https://doi.org/10.1145/1963405.1963487>`_
+
+    .. [6] `Leland McInnes
+        "PyNNDescent for fast Approximate Nearest Neighbors"
+        <https://pynndescent.readthedocs.io/en/latest/>`_
+    """
+    train_size, valid_size, test_size = validate_train_valid_test_split_sizes(
+        train_size, valid_size, test_size, len(data)
+    )
+
+    clusters = _create_clusters(data, threshold, approximate, n_jobs)
+    clusters.sort(key=len)
+
+    train_idxs: list[int] = []
+    valid_idxs: list[int] = []
+    test_idxs: list[int] = []
+
+    for cluster in clusters:
+        if len(test_idxs) < test_size:
+            test_idxs.extend(cluster)
+        elif len(valid_idxs) < valid_size:
+            valid_idxs.extend(cluster)
+        else:
+            train_idxs.extend(cluster)
+
+    if return_indices:
+        train_subset = train_idxs
+        valid_subset = valid_idxs
+        test_subset = test_idxs
+    else:
+        train_subset = get_data_from_indices(data, train_idxs)
+        valid_subset = get_data_from_indices(data, valid_idxs)
+        test_subset = get_data_from_indices(data, test_idxs)
+
+    ensure_nonempty_subset(train_subset, "train")
+    ensure_nonempty_subset(valid_subset, "validation")
+    ensure_nonempty_subset(test_subset, "test")
+
+    if additional_data:
+        additional_data_split: list[Sequence[Any]] = split_additional_data(
+            list(additional_data), train_idxs, valid_idxs, test_idxs
+        )
+        return train_subset, valid_subset, test_subset, *additional_data_split
+    else:
+        return train_subset, valid_subset, test_subset
+
+
+def _create_clusters(
+    data: Sequence[Union[str, Mol]],
+    threshold: float = 0.65,
+    approximate: bool = False,
+    n_jobs: Optional[int] = None,
+) -> list[list[int]]:
+    """
+    Generate Taylor-Butina clusters for a list of SMILES strings or RDKit `Mol` objects.
+    This function groups molecules by using clustering, where cluster centers must have
+    Tanimoto (Jaccard) distance greater or equal to given threshold. Binary ECFP4 (Morgan)
+    fingerprints with 2048 bits are used as features.
+    """
+    mols = ensure_mols(data)
+
+    fps_rdkit = GetMorganGenerator().GetFingerprints(mols)
+    centroid_idxs = LeaderPicker().LazyBitVectorPick(
+        fps_rdkit, poolSize=len(mols), threshold=threshold
+    )
+    centroid_idxs = list(centroid_idxs)
+    non_centroid_idxs = sorted(set(range(len(mols))) - set(centroid_idxs))
```

Comment:
GetFingerprints returns a tuple of `ExplicitBitVect`. Wouldn't it be easier to convert the results yielded by the skfp transformer to `ExplicitBitVect` instead of converting the datasets with Morgan generator?

Reply:
No, `ExplicitBitVect` is a C++ object and cannot be created from Python. This is the only way we can do this, as far as I know